opennms-web-api/src/main/java/org/opennms/web/enlinkd/EnLinkdElementFactory.java

```diff
@@ -995,43 +1024,94 @@ private String getIpList(Set<String> ipstrings) {
+                        return Pair.of(list, uniqueMapCache);
+                    }
+            );
+        }
+
+        public List<T> get(int nodeId) {
+            return pair(nodeId).getLeft();
+        }
+
+        public T get(int nodeId, S selection) {
+            if (selection == null) {
                 return null;
             } else {
-                return map.computeIfAbsent(nodeId, n ->
-                        uniqueMap(m_snmpInterfaceDao.findByNodeId(n), OnmsSnmpInterface::getIfIndex, OnmsSnmpInterface::getId)
-                ).get(ifIdx);
+                return pair(nodeId).getRight().get(selection);
             }
         }
+
     }
 
-    private class BridgeElementCache {
-        private Map<Integer, Map<Integer, BridgeElement>> map = new HashMap<>();
-        public BridgeElement get(int nodeId, Integer vlan) {
-            if (vlan == null) {
-                return null;
-            } else {
-                return map.computeIfAbsent(nodeId, n ->
-                        uniqueMap(m_bridgeElementDao.findByNodeId(n), BridgeElement::getVlan, BridgeElement::getId)
-                ).get(vlan);
-            }
+    private class SnmpInterfaceCache extends ListAndUniqueMapCache<OnmsSnmpInterface, Integer> {
+        public SnmpInterfaceCache() {
+            super(m_snmpInterfaceDao::findByNodeId, OnmsSnmpInterface::getIfIndex, OnmsSnmpInterface::getId);
+        }
+    }
+
+    private class BridgeElementCache extends ListAndUniqueMapCache<BridgeElement, Integer> {
+        public BridgeElementCache() {
+            super(m_bridgeElementDao::findByNodeId, BridgeElement::getVlan, BridgeElement::getId);
         }
     }
 
     /**
-     * Creates a map that contains entries for those keys that have a unique value.
+     * Like a map but may be initialized lazily.
+     */
+    private interface UniqueMapCache<K, V> {
+        V get(K key);
+    }
+
+    /**
+     * Creates a UniqueCacheMap. The map is loaded lazily.
+     */
+    private static <K, V, I> UniqueMapCache<K, V> uniqueMapCache(Supplier<List<V>> supplier, Function<V, K> toKey, Function<V, I> toId) {
+        return new UniqueMapCache<>() {
+            private Map<K, Optional<V>> map;
+            public V get(K key) {
+                if (map == null) {
+                    map = uniqueMap(supplier.get(), toKey, toId);
+                }
+                var o = map.get(key);
+                return o == null ? null : o.orElse(null);
+            }
+        };
+    }
+
+    /**
+     * Creates a map that contains entries for all keys derived from the values. Keys are mapped to a non-empty optionals
+     * in case the mapping is unique and to an empty optional otherwise.
      */
-    private static <K, V, I> Map<K, V> uniqueMap(List<V> list, Function<V, K> toKey, Function<V, I> toId) {
+    private static <K, V, I> Map<K, Optional<V>> uniqueMap(List<V> list, Function<V, K> toKey, Function<V, I> toId) {
```

Comment:
This usage of `Optional` looks strange. Is there a distinction between a missing value and am explicitly empty value?

Only left usage seems to be the call above, which results in a `o == null ? null : o.orElse(null)`. Which makes this even more strange.

Reply:
The `toMap` collector does not distinguish between keys that are new and keys that have a `null` value. In both case the next value is added to the collected map without using the combine function. Therefore `null` can not be used as the `value` for indicating a key-value mapping is not unique.